crates/leon/src/values.rs

```diff
@@ -0,0 +1,209 @@
+}
+
+impl Values for Vec<(&str, &str)> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == &key {
+                Some(Cow::Borrowed(*v))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl Values for HashMap<&str, &str> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.get(&key).map(|v| Cow::Borrowed(*v))
+    }
+}
+
+impl Values for BTreeMap<&str, &str> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.get(&key).map(|v| Cow::Borrowed(*v))
+    }
+}
+
+impl Values for [(String, &str)] {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == key {
+                Some(Cow::Borrowed(*v))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl<const N: usize> Values for [(String, &str); N] {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == key {
+                Some(Cow::Borrowed(*v))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl Values for Vec<(String, &str)> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == key {
+                Some(Cow::Borrowed(*v))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl Values for HashMap<String, &str> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.get(key).map(|v| Cow::Borrowed(*v))
+    }
+}
+
+impl Values for BTreeMap<String, &str> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.get(key).map(|v| Cow::Borrowed(*v))
+    }
+}
+
+impl Values for [(String, String)] {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == key {
+                Some(Cow::Owned(v.clone()))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl<const N: usize> Values for [(String, String); N] {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == key {
+                Some(Cow::Owned(v.clone()))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl Values for Vec<(String, String)> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == key {
+                Some(Cow::Owned(v.clone()))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl Values for HashMap<String, String> {
```

Comment:
If we have impl for both `HashMap<String, &str>` and `HashMap<String, String>`, then does it make sense to have impl for `HashMap<&str, String>` and `HashMap<&str, &str>`?

Same for `BTreeMap`.

Also, there's a lot of dup impl for `HashMap` and `BTreeMap`, I wish we can dedup them to reduce codegen or use a `macro_rules!` to at least make maintaining this easier.

Reply:
For `HashMap`, we could do:

```rust
impl<K, V> for HashMap<K, V>
where
    K: Borrow<str>,
    V: AsRef<str>
```